docs/modules/SDDs/pages/0021-cluster-catalog-compilation.adoc

```diff
@@ -0,0 +1,170 @@
+By using the Kapitan inventory the tool needs to figure out for which clusters the config changed and therefore needs a recompilation.
+In a first step this could be relatively rudimentary and only use the input information from a webhook (cluster/tenant ID) to decide which clusters to compile.
+
+The information if a cluster needs to be compiled is then stored in the cluster object of the Lieutenant API.
+In a first PoC phase this can be an annotation like `cluster.syn.tools/needs-compilation=true`.
+In a later stage this information should be part of the typed struct and might include further information like the exact version of the various input repos.
+
+
+==== Inventory Parsing
+
+In order for Quartermaster to figure out which clusters are affected by a certain config change, the Kapitan inventory needs to be parsed.
+Kapitan does this by using https://github.com/salt-formulas/reclass[reclass] which is a tool for merging and searching data sources recursively in YAML files.
+
+Quartermaster can use reclass to parse the inventory and search for clusters that are affected by changes in a certain class.
+One issue we've to work around in this approach is, that reclass can only parse the inventory if it's complete and all referenced classes are available.
+This is a problem since creating the whole inventory (with all components) would be too resource and time intensive.
+As a workaround Quartermaster should only have to get the global and tenant parts of the inventory.
+The classes for the different Commodore components should be faked and generated as empty files.
+The only downside to this approach is that if a component class imports another class, this won't be captured by Quartermaster which might result in cluster catalogs not being compiled.
+Since we currently don't allow components to include classes this isn't a problem.
+
+
+==== Webhooks
+
+Triggering a webhook can be implemented in either a Git repo natively (for example https://developer.github.com/webhooks/[GitHub Webhooks]) or via a CI/CD system (for example a https://docs.gitlab.com/ee/ci/[GitLab CI] job which runs `curl`).
+A webhook can also be called by any other tool or even manually with a `curl` command.
+
+A webhook can include the following optional information:
+
+* Cluster ID for which to compile the catalog.
+This can be used in case where a specific cluster should be compiled
+
+* Tenant ID for which to figure out affected clusters.
+This can be used to limit the set of clusters to be considered
+
+* Set of classes that changed
+
+
+==== Authentication & Authorization
+
+To prevent exsessive resource usage or even DoS attacks, the webhooks need to be authenticated.
+A permission system needs to be in place to decide which entities are allowed to trigger which actions.
+It shouldn't be possible for example for a tenant to trigger compilation for clusters of other tenants.
+
```

Comment:
Since we are talking about webhooks, this kind of authentication must work as part of a GET request and thus has to be part of the URL.

Reply:
GitHub as well as GitLab do `POST` requests by default for webhooks.

Reply:
But still, you can not configure the `Authorization` header. Github allows you do define a _secret_ which is then submitted in the header  `X-Hub-Signature`.

I phrase this as a more general question: how do we want authentication to work?